Dynamic Movement Primitives Implementation

SchwartzCode opened the pull request:
#### What does this implement/fix?
Implements [Dynamic Movement Primitives](https://www.frontiersin.org/articles/10.3389/fncom.2013.00138/full), which is a way of 'learning' a path and being able to stretch or squeeze it in space or time as the user sees fit. The path is learned as a weighted sums of gaussian distributions.

#### Additional information
I made a file and a few unit tests, but let me know if there is anything else you'd like me to add to this!

SchwartzCode reviewed (commented):
Looks good to me (but I also wrote this lol so of course it looks good to me). Only thing I didn't add was an entry to the table of contents in the README, but I figured I'd wait on that to see if this PR is approved :)

SchwartzCode commented:
Also, all commits before #d5b4361 were merged in with the two length=0 fixes I pushed for dubins planner and reeds shepp planners last week, not really sure why they are showing in this PR too

AtsushiSakai commented:
Thank you for your PR again. Is it possible to add an arm navigation simulation using this technique?

SchwartzCode commented:
I think so, do you want me to add those changes to this PR or should I close and re-make it once I've finished adding things?

AtsushiSakai commented:
It is ok to add it in this PR.

SchwartzCode commented:
@AtsushiSakai passing everything except CodeFactor, which is mad about some for loops in `ArmNavigation\n_joint_arm_3d\NLinkArm3d.py`, let me know if there's any other changes to this you'd like!

AtsushiSakai commented on PathPlanning/DynamicMovementPrimitives/dynamic_movement_primitives.py:
```diff
@@ -0,0 +1,260 @@
+        self.weights = None  # weights used to generate DMP trajectories
+
+        self.T_orig = data_period
+        # self.training_data = training_data
```
This line should be removed?

AtsushiSakai commented on PathPlanning/DynamicMovementPrimitives/dynamic_movement_primitives.py:
```diff
@@ -0,0 +1,260 @@
+        H = (0.65*(1./(nrBasis-1))**2)
+
+        # initialize virtual system
+        t = 0
```
`time` is better?

AtsushiSakai commented on PathPlanning/DynamicMovementPrimitives/dynamic_movement_primitives.py:
```diff
@@ -0,0 +1,260 @@
+        """
+        init_state - initial state/position
+        goal_state - goal state/position
+        T  - amount of time to travek q0 -> g
```
`travek` -> `travel`?

AtsushiSakai commented on PathPlanning/DynamicMovementPrimitives/dynamic_movement_primitives.py:
```diff
@@ -0,0 +1,260 @@
+
+if __name__ == '__main__':
+
+    t = np.arange(0, 3*np.pi/2, 0.01)
```
This example codes should be in a function `main` or something I think.

SchwartzCode replied:
Added a function example_DMP() that does everything that was originally inside this conditional

AtsushiSakai commented on tests/test_dynamic_movement_primitives.py:
```diff
@@ -0,0 +1,35 @@
+    train_data = np.array([t, sin_t]).T
+
+    DMP_controller = dynamic_movement_primitives.DMP(train_data, T)
+    DMP_controller.recreate_trajectory(train_data[0], train_data[-1], 4)
```
Maybe a test to check the path result is not empty here?

SchwartzCode replied:
Added a test to do this and named it test 2 (old test 2 that checks if path is close to training data is now test 3)

AtsushiSakai commented on PathPlanning/DynamicMovementPrimitives/dynamic_movement_primitives.py:
```diff
@@ -0,0 +1,260 @@
+
+        for new_q0_value in q0_vals:
+            plot_title = "Initial Position = [" + \
+                         str(round(new_q0_value[0], 2)) + ", " \
```
I think f-string makes these codes cleaner.

SchwartzCode replied:
Definitely! Updated

AtsushiSakai commented on PathPlanning/DynamicMovementPrimitives/dynamic_movement_primitives.py:
```diff
@@ -0,0 +1,260 @@
+            for dim in range(dimensions):
+
+                if t <= T:
+                    Phi = [np.exp(-0.5 * ((t / T) - c)**2 / H) for c in C]
```
`Phi` -> `phi` is better?

AtsushiSakai commented on PathPlanning/DynamicMovementPrimitives/dynamic_movement_primitives.py:
```diff
@@ -0,0 +1,260 @@
+            qd = qd + qdd * self.dt
+            q = q + qd * self.dt
+
+            if not isinstance(q, list):
```
`q` is not always `np.array`? If so I think the instance check is redundant.

SchwartzCode replied:
It would be on the first loop, I updated this function so everything is numpy arrays

AtsushiSakai commented on PathPlanning/DynamicMovementPrimitives/dynamic_movement_primitives.py:
```diff
@@ -0,0 +1,260 @@
+"""
```
Please add a one-line comment to describe this example here.

SchwartzCode replied:
Added one to the example_DMP() function toward the bottom of the file

AtsushiSakai replied:
Can you add a one-line comment to describe this file here? Like: "Dynamic Movement Primitives examples for robotics".

AtsushiSakai reviewed (changes requested):
Thank you for your work. I have some comments. PTAL.

SchwartzCode commented:
@AtsushiSakai addressed most of your comments (resolved ones where I just did what was suggested, left ones open that I maybe didn't do exactly what was asked), let me know if there are more changes you'd like to see :)

AtsushiSakai commented on PathPlanning/DynamicMovementPrimitives/dynamic_movement_primitives.py:
```diff
@@ -0,0 +1,257 @@
+        time = 0
+
+        # for plotting
+        self.train_t_vals = np.arange(0, T, self.timesteps)
```
Where is this used for plotting?

SchwartzCode replied:
Nowhere, deleted!

AtsushiSakai reviewed (changes requested):
Thank you. This is the last review. After, addressed these, I will merge it.

SchwartzCode commented:
@AtsushiSakai made the changes you requested, thanks for making this repo! It helped me a ton this summer while prototyping path planners :)

AtsushiSakai reviewed (approved):
LGTM. I will merge it. Thank you for your work!! @SchwartzCode
